_posts/2024-08-04-non-academic-problem.md

```diff
@@ -0,0 +1,134 @@
+
+We can keep count of such chords, and when a vertex is finished, we can determine whether the edge between it and its parent is a bridge.
+
+![bridge](/assets/posts/2024-08-04-non-academic-problem/chords.svg)
+
+Here, you can see a possible search tree starting from node $1$. The dashed lines mark chords.
+
+To find possible soulutions, we can run a second DFS to count the vertices in the subtree of each bridge. Let the number of vertices be $n$ in the whole graph, and $k$ in the current subtree. Then the graph has $\binom{n}{2}$ paths in total, and we can break $k * (n - k)$ of those by removing the current edge.
+
+The complexity of this method is $O(n + m)$ since we only visit the vertices and edges once in each DFS.
+
+## Implementation
+
+The below implementation involves a recursive version of DFS, extended to account for chords. A chord is considered finished, if both of its endpoints are finished. We determine an edge $(u, v)$ to be a bridge, if and only if there are no unfinished chords remaining in $v$'s subtree, after $v$ has been finished.
+
+In the second DFS, we count the vertices of the subtrees, and choose between the possible solutions at the bridges.
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using vi = vector<int>;
+using ll = long long;
+#define ff(i, n) for(int i = 0; i < (int)(n); ++i)
+
+
+struct edge
+{
+    int t; // target node
+    bool bridge;
+};
+struct node
+{
+    vector<edge> adj;
+
+    bool vis;
+    int chor;
+};
+
+int dfs_bridge(vector<node> &g, int p, int i)
+{
+    node &v = g[i];
+    if(v.vis) { ++v.chor; return 1; } // v.chor stores the number of chords that end in v
+    v.vis = true;
+
+    int chor = 0; // count of the unfinished chords in the subgraph
+    for(edge &e : v.adj)
+    {
+        if(e.t == p) continue;
+
+        int r = dfs_bridge(g, i, e.t);
+        chor += r;
+        e.bridge = (r == 0); 
+    }
+    // return the number of unfinished chords
+    return chor - v.chor * 2; // chords that end in v were counted from both sides
+}
+
+int dfs_sol(vector<node> &g, int p, int i, ll &best)
+{
+    node &v = g[i];
+    if(v.vis) { return 0; }
+    v.vis = true;
+
+    int count = 1;
+    for(edge &e : v.adj)
+    {
+        if(e.t == p) continue;
+
+        int r = dfs_sol(g, i, e.t, best);
+        if(e.bridge)
+            // best stores the largest number of paths that can be broken
+            best = max(best, ((ll)g.size() - r) * r);
+        count += r;
+    }
+    // return the number of vertices in the subgraph
+    return count;
+}
+
+int main()
+{
+    int t; cin >> t; while(t--)
+    {
+        int n, m; cin >> n >> m;
+        vector<node> g(n, (node){ vector<edge>(), false, 0 });
```

Comment:
@leventegego

```suggestion
        vector<node> g(n);
```
Is this for readability? This calls the default constructor of node too.